src/app/search-page/search.component.ts

```diff
@@ -0,0 +1,226 @@
+import { Subject, takeUntil } from 'rxjs';
+import { ProductService } from '../api/products.service';
+import { QueryService } from '../api/query.service';
+import { IProduct } from '../Interface/products';
+import { FilterService } from '../services/filter.services';
+
+
+@Component({
+  selector: 'app-search',
+  templateUrl: './search.component.html',
+  styleUrls: ['./search.component.scss'],
+})
+
+export class SearchComponent implements OnInit, OnDestroy {
+
+  query: string | null = '';
+
+  allProductsBySearch: IProduct[] = [];
+
+  filteredProducts: IProduct[] = [];
+
+  searchCount: number | undefined;
+
+  visibleProducts: IProduct[] = [];
+
+  categoryList: string[] = [];
+
+  statusList: string[] = [];
+
+  selectedCategory: string = '';
+
+  selectedStatus: string = '';
+
+  selectedReleaseDate: string = '';
+
+  private readonly destroy$ = new Subject<void>();
+
+  private _searchDetails: string = '';
+
+  constructor(private route: ActivatedRoute,
+    private router: Router,
+    private productService: ProductService,
+    private queryService: QueryService,
+    private filterService: FilterService) { }
+
+
+  get searchDetails(): string {
+    return this._searchDetails;
+  }
+
+  set searchDetails(value: string) {
+    this._searchDetails = value;
+  }
+
+  resetFilterValues() {
+    this.selectedStatus = '';
+    this.selectedCategory = '';
+    this.selectedReleaseDate = '';
+  }
+
+  onSearch(): void {
+    this.router.navigate(["search", this._searchDetails]);
+    this.query = this._searchDetails;
+
+    this.productService.getProductsByQuery(this._searchDetails)
+      .pipe(takeUntil(this.destroy$))
+      .subscribe({
+        next: data => {
+          this.resetFilterValues();
+          this.filteredProducts = [];
+          this.allProductsBySearch = data.results;
+          this.searchCount = data.count;
+          this.visibleProducts = this.allProductsBySearch.slice(0, 12);
+          this.getCategoryList();
+          this.getStatusList();
+        }
+      })
+  }
+
+  onPageChange(event: PageEvent, productsArray: IProduct[]) {
+
+    const startIndex = event.pageIndex * event.pageSize;
+    let endIndex = startIndex + event.pageSize;
+
+    if (endIndex > productsArray.length) {
+      endIndex = productsArray.length
+    }
+    this.visibleProducts = productsArray.slice(startIndex, endIndex);
+
+  }
+
+  getCategoryList() {
+    const unfilteredCategories: string[] = [];
+
+    this.allProductsBySearch.map(item => {
+      unfilteredCategories.push(item.category)
+    })
+    const uniqueCategories = findUniqueItems(unfilteredCategories);
+
+    this.categoryList = uniqueCategories;
+  }
+
+  getStatusList() {
+    const unfilteredStatusList: string[] = [];
+
+    this.allProductsBySearch.map(item => {
+      unfilteredStatusList.push(item.status)
+    })
+    const uniqueStatusList = findUniqueItems(unfilteredStatusList);
+
+    this.statusList = uniqueStatusList;
+  }
```

Comment:
These functions are almost identical.
When this is the case it makes sense to move the identical parts into a new function.

Reply:
- [x] done: created one function getFilterList()